toolchain/check/context.cpp

```diff
@@ -799,30 +801,79 @@ auto Context::SetBlockArgResultBeforeConstantUse(SemIR::InstId select_id,
   }
 }
 
-auto Context::AddCurrentCodeBlockToFunction(Parse::NodeId node_id) -> void {
-  CARBON_CHECK(!inst_block_stack().empty(), "no current code block");
-
-  if (return_scope_stack().empty()) {
-    CARBON_CHECK(node_id.is_valid(),
-                 "No current function, but node_id not provided");
-    TODO(node_id,
-         "Control flow expressions are currently only supported inside "
-         "functions.");
+auto Context::AddToRegion(SemIR::InstBlockId block_id) -> void {
+  if (region_stack_.empty() || block_id == SemIR::InstBlockId::Unreachable) {
     return;
   }
 
-  if (!inst_block_stack().is_current_block_reachable()) {
-    // Don't include unreachable blocks in the function.
-    return;
+  region_stack_.back().block_ids.push_back(block_id);
+}
+
+auto Context::BeginSubpattern() -> void {
+  auto entry_block_id = sem_ir().inst_blocks().AddDefaultValue();
+  inst_block_stack().Push(entry_block_id);
+  PushRegion(entry_block_id);
+}
+
+auto Context::EndSubpatternAsExpression(SemIR::InstId result_id)
+    -> SemIR::RegionId {
+  // TODO: Is it possible to validate that this region is genuinely
+  // single-entry, single-exit?
+  SemIR::Region region = PopRegion(result_id);
+  CARBON_CHECK(region.block_ids.back() == inst_block_stack().PeekOrAdd());
+  if (region.block_ids.size() > 1) {
+    // End the exit block with a branch to a successor block, whose contents
+    // will be determined later.
+    AddInst(SemIR::LocIdAndInst::NoLoc<SemIR::Branch>(
+        {.target_id = inst_blocks().AddDefaultValue()}));
+  } else {
+    // This single-block region will be inserted as a SpliceBlock, so we don't
+    // need control flow out of it.
   }
+  inst_block_stack().Pop();
+  return sem_ir().regions().Add(region);
+}
+
+auto Context::EndSubpatternAsEmpty() -> void {
+  auto region = region_stack_.pop_back_val();
+  CARBON_CHECK(region.block_ids.front() == inst_block_stack().PeekOrAdd());
+  CARBON_CHECK(inst_block_stack().PeekCurrentBlockContents().empty());
+  inst_block_stack().PopAndDiscard();
```

Comment:
Maybe:
```suggestion
  CARBON_CHECK(inst_block_stack().PeekCurrentBlockContents().empty());
  auto exit_block_id = inst_block_stack().Pop();
  CARBON_CHECK(region.block_ids.front() == exit_block_id);
```
The `PopAndDiscard` is misleading given that the block was definitely already allocated an `InstBlockId`, and this avoids the appearance of a side-effect in a `CHECK`.

Reply:
How about this instead? I don't want to reverse the order of the checks because in some circumstances, "this isn't the block you expected" could be much more enlightening than "this block doesn't have the contents you expect".